licenses/library.go

```diff
@@ -169,35 +172,48 @@ func (l *Library) String() string {
 	return l.Name()
 }
 
-// FileURL attempts to determine the URL for a file in this library.
-// This only works for certain supported package prefixes, such as github.com,
-// bitbucket.org and googlesource.com. Prefer GitRepo.FileURL() if possible.
-func (l *Library) FileURL(filePath string) (*url.URL, error) {
-	relFilePath, err := filepath.Rel(filepath.Dir(l.LicensePath), filePath)
-	if err != nil {
-		return nil, err
+// FileURL attempts to determine the URL for a file in this library using
+// go module name and version.
+func (l *Library) FileURL(ctx context.Context, filePath string) (string, error) {
+	if l == nil {
+		return "", fmt.Errorf("library is nil")
 	}
-	nameParts := strings.SplitN(l.Name(), "/", 4)
-	if len(nameParts) < 3 {
-		return nil, fmt.Errorf("cannot determine URL for %q package", l.Name())
+	wrap := func(err error) error {
+		return fmt.Errorf("getting file URL in library %s: %w", l.Name(), err)
 	}
-	host, user, project := nameParts[0], nameParts[1], nameParts[2]
-	pathPrefix, ok := repoPathPrefixes[host]
-	if !ok {
-		return nil, fmt.Errorf("unsupported package host %q for %q", host, l.Name())
+	m := l.Module
+	if m == nil {
+		return "", wrap(fmt.Errorf("empty go module info"))
 	}
-	if len(nameParts) == 4 {
-		pathPrefix = path.Join(pathPrefix, nameParts[3])
+	if m.Dir == "" {
+		return "", wrap(fmt.Errorf("empty go module dir"))
 	}
-	return &url.URL{
-		Scheme: "https",
-		Host:   host,
-		Path:   path.Join(user, project, pathPrefix, relFilePath),
-	}, nil
+	client := source.NewClient(time.Second * 20)
+	ver := m.Version
+	if ver == "" {
+		// This always happens for the module in development.
+		ver = "master"
+		glog.Warningf("module %s has empty version, defaults to master. The file URL may be incorrect. Please verify!", m.Path)
+	}
+	remote, err := source.ModuleInfo(ctx, client, m.Path, ver)
+	if err != nil {
+		return "", wrap(err)
+	}
+	relativePath, err := filepath.Rel(m.Dir, filePath)
+	if err != nil {
+		return "", wrap(err)
+	}
+	// TODO: there are still rare cases this may result in an incorrect URL.
+	// https://github.com/google/go-licenses/issues/73#issuecomment-1005587408
```

Comment:
Might be worthwhile to document examples of the types of URLs that are problematic (I didn't get a clear sense what "major branch conventions" meant from the comment)

Reply:
Sure, let me expand on the comment and leave a TODO in v2 roadmap to document known caveats.

Reply:
Updated in https://github.com/google/go-licenses/issues/73#issuecomment-1005587408, note I have already built workarounds for these special cases in follow up PRs.